cli/sync_service/download.go

```diff
@@ -109,23 +90,49 @@ func ObjectDownLoad(org string, userPw string, objType string, objId string, fil
 		}
 	}
 
-	file, err := os.Create(fileName)
-	if err != nil {
-		cliutils.Fatal(cliutils.INTERNAL_ERROR, msgPrinter.Sprintf("Failed to create file: %s", fileName))
-	}
+	rename := false
+	if !skipDigitalSigVerify {
+		// Verify digital signature, save to a tmp file, rename tmp file
+		// Call the MMS service over HTTP to get object metadata
+		var objectMeta common.MetaData
+		urlPath = path.Join("api/v1/objects/", org, objType, objId)
+		httpCode := cliutils.ExchangeGet("Model Management Service", cliutils.GetMMSUrl(), urlPath, cliutils.OrgAndCreds(org, userPw), []int{200, 404}, &objectMeta)
+		if httpCode == 404 {
+			cliutils.Fatal(cliutils.NOT_FOUND, msgPrinter.Sprintf("object metadata '%s' of type '%s' not found in org %s", objId, objType, org))
+		}
 
-	defer file.Close()
+		if objectMeta.HashAlgorithm != "" && objectMeta.PublicKey != "" && objectMeta.Signature != "" {
+			// verify data
+			//dataReader := bytes.NewReader(data)
+			msgPrinter.Println("Verifying data with digital signature....")
+			if verified, err := VerifyDataSig(data, objectMeta.PublicKey, objectMeta.Signature, objectMeta.HashAlgorithm, fileName); !verified {
+				cliutils.Fatal(cliutils.INTERNAL_ERROR, msgPrinter.Sprintf("Failed to verify data: %s", err.Error()))
+			}
+			msgPrinter.Println("Verifying digital signature is done.")
+			rename = true
+		}
 
-	if _, err := file.Write(data); err != nil {
-		cliutils.Fatal(cliutils.INTERNAL_ERROR, msgPrinter.Sprintf("Failed to save data for object '%s' of type '%s' to file %s", objId, objType, fileName))
+	}
+
+	if rename {
+		// rename fileName.tmp
+                tmpFileName := fmt.Sprintf("%s.tmp", fileName)
+                if err := os.Rename(tmpFileName, fileName); err != nil {
+                        cliutils.Fatal(cliutils.INTERNAL_ERROR, msgPrinter.Sprintf("Failed to rename file '%s' to '%s', err: %v", tmpFileName, fileName, err))
+                }
+	} else {
```

Comment:
This else clause would become the else of the if !skipDigitalVerify

Reply:
Yes, that is right. So 
if !skipDigitalVerify, 
     the data will be streamed to a tmp file, at same time generate hash + verify.
then rename the tmp file
else 
    the data is stream to the file directly

Reply:
@dabooz ah I think I know what you mean. Yea after moving the rename logic, will also change this part